app/src/main/java/se/nullable/flickboard/KeyboardService.kt

```diff
@@ -402,11 +404,26 @@ class KeyboardService : InputMethodService(), LifecycleOwner, SavedStateRegistry
 
                                 Action.ToggleEmojiMode ->
                                     emojiMode = !emojiMode
+
+                                Action.EnableVoiceMode -> {
+                                    voiceMode = true
+                                }
                             }
                         }
                         Box {
                             when {
                                 emojiMode -> EmojiKeyboard(onAction = onAction)
+                                voiceMode -> {
+                                    voiceMode = false
+                                    val voiceMethodId = getAvailableVoiceInputMethod()
+
+                                    if(voiceMethodId != null) {
+                                        // Only works on Android 9+
```

Comment:
We only support Android 9+ anyway. But we still need a fallback for when the user has no voice.